lib/FusionInventory/Agent/Task/Inventory/BSD/Storages.pm

```diff
@@ -17,39 +19,111 @@ sub doInventory {
     my $inventory = $params{inventory};
     my $logger    = $params{logger};
 
-    # get a list of devices from /etc/fstab
-    my @devices = _getDevicesFromFstab(logger => $logger);
+    my $storages = _retrieveStoragesFromSysCtl(
+        logger => $logger
+    );
+    for my $storage (@$storages) {
+        $inventory->addEntry(
+            section => 'STORAGES',
+            entry   => $storage
+        );
+    }
+}
+
+sub _retrieveStoragesFromSysCtl {
+    my (%params) = @_;
 
-    # parse dmesg
-    my @lines = getAllLines(
-        command => 'dmesg'
+    my @storages = _getDevicesFromSysCtl(
+        %params,
+        file => $params{sysctlFile} ? $params{sysctlFile} : ''
+    );
+    _extractDataFromDmesg(
+        file => $params{dmesgFile} ? $params{dmesgFile} : '',
+        storages => \@storages
     );
 
-    foreach my $device (@devices) {
+    return @storages;
+}
 
-        foreach my $line (@lines) {
-            if ($line =~ /^$device->{DESCRIPTION}.*<(.*)>/) {
-                $device->{MODEL} = $1;
-            }
-            if ($line =~ /^$device->{DESCRIPTION}.*\s+(\d+)\s*MB/) {
-                $device->{CAPACITY} = $1;
-            }
+sub _getDevicesFromSysCtl {
+    my (%params) = @_;
+
+    my $command = 'sysctl kern.geom.confxml';
+    my $lines = getAllLines(
+        command => $command,
+        %params
+    );
+    $lines =~ s/^kern.geom.confxml://;
+    my $tpp = XML::TreePP->new();
+    my $tree = $tpp->parse($lines);
+
+    my @devices = ();
+    for my $class (@{$tree->{mesh}->{class}}) {
+        my $name = $class->{name} || $class->{'#name'} || '';
+        print 'name: ' . $name . "\n";
+        next unless $name && $name eq 'DISK';
```

Comment:
For "unless" and following "if" statements, can you group tested conditions into parenthesis while you check more than 1 condition ? This helps reading the code.